### ISSUE TITLE
Performance issues when parsing null conditional operators in C#

### ISSUE THREAD

**Issue by Sjord** [semgrep/semgrep#6018]
This file takes several minutes to parse:

```
public class Foo
{
    private static bool MyFunction(object x)
    {
        return 
            x.A?.B 
            || x.A?.C 
            || x.D?.E 
            || x.D?.F 
            || x.G?.B 
            || x.G?.H 
            || x.G?.I 
            || x.G?.J 
            || x.G?.K 
            || x.G?.L 
            || x.G?.C 
            || x.G?.N?.B 
            || x.G?.M?.B 
            || x.G?.O?.B 
            || x.G?.P?.B 
            || x.Q?.B ;
    }
}
```

```
$ time bin/semgrep-core -dump_ast -lang csharp loop.cs > /dev/null
bin/semgrep-core -dump_ast -lang csharp loop.cs > /dev/null  102.77s user 29.71s system 79% cpu 2:46.45 total
```

```
$ time semgrep -l csharp -e '$A=3' loop.cs
Scanning 1 file.
Warning: 1 timeout error(s) in loop.cs when running the following rules: [-]
```

**Comment by r2c-demo** [comment]
This issue is synced in Linear at https://linear.app/r2c/issue/PA-1837/performance-issues-when-parsing-null-conditional-operators-in-c. _Note_: this link is for r2c use only and is not accessible publicly.

**Comment by Sjord** [comment]
There are two things causing this issue:

* [Precedence incorrect between or and conditional access](https://github.com/tree-sitter/tree-sitter-c-sharp/issues/233)
* map `a?.b` to `null == a ? null : a.b` [here](https://github.com/returntocorp/semgrep/blob/756408aafebb849d89993a067f9c6f8d834df8a4/semgrep-core/src/parsing/tree_sitter/Parse_csharp_tree_sitter.ml#L1295)

So `x || x?.C` is parsed as `(x || x)?.C`, and that first expression is used twice in the conversion to `null == (x || x) ? null : (x || x).C`. With 20 conditional expressions, we get 2²⁰ or about a million entries in the syntax tree. And that's slow.

So this can be solved by fixing the precedence in tree-sitter-c-sharp, but having a conditional [DotAccess](https://github.com/returntocorp/semgrep/blob/756408aafebb849d89993a067f9c6f8d834df8a4/semgrep-core/src/core/ast/AST_generic.ml#L508) in AST_Generic would also make sense.

**Comment by Sjord** [comment]
I am getting pretty tired of the stale bot.

**Comment by aryx** [comment]
cc @brandonspark

**Comment by Sjord** [comment]
Could you reopen this please?

**Comment by aryx** [comment]
One possible fix is to use an intermediate variable to store the value of the expression a, so we don't repeat
a each time so 
we map `a?.b` to `let va = a in null == va ? null : va.b`

**Comment by aryx** [comment]
This should avoid the AST explosion.

**Comment by aryx** [comment]
@IagoAbal could we do this rewriting? (introduce some intermediate var in the generic AST). Would this interfere
with constant propagation and matching? or Let is already handled there?

**Comment by IagoAbal** [comment]
Could we instead do a translation based on the `Elvis` operator as @brandonspark did here: https://github.com/returntocorp/semgrep/pull/6265/files ? Basically `x?.y` becomes `(x?).y`.

**Comment by aryx** [comment]
sounds good. will try that monday.

**Comment by Sjord** [comment]
> Basically x?.y becomes (x?).y

This is semantically incorrect, especially with function calls. Consider `x?.y()`. In `(x?).y()`, `y()` is always called, possibly on a null value. When in reality, the function call is not performed when x is null.

In my opinion the `?.` needs it's own operator in AST_Generic. It is pretty common in multiple languages, and is sufficiently different from `.`. It cannot be mapped to something else without introducing other problems. And currently it is not possible to match the `?.` operator.

**Comment by aryx** [comment]
I see. You're right it's incorrect, but semgrep is not trying to be a syntax checker/type checker/interpreter, so for matching purpose, the transformation is fine. The only issue is that if you look for $O.$M(), then this would match too (x?).y(), but maybe that's actually something we want.

**Comment by Sjord** [comment]
In that case, perhaps it's simpler to just map `a?.b` to `a.b`. That loses the distinction between the two, but as you said that might not be such a big problem.

> Basically x?.y becomes (x?).y

> G.Call (G.IdSpecial (G.Op G.Elvis, t) |> G.e, fb [ G.Arg e ])

This makes no sense to me. I thought Elvis needs two operands. E.g. `a ?: b` returns `b` if `a` is null and `a` otherwise. But in this code it seems to be called with one operand. If `x?.y` becomes `(x?).y`, what does `x?` even mean in this example?

**Comment by IagoAbal** [comment]
Internally, for Semgrep, `(x?).y` means `x?.y`. This is not a new thing, we abuse other Generic AST constructs in a similar way already. If we didn't do this kind of thing then Generic would probably have to grow considerably. (To be fair, I'm speculating a bit here, since I have not been personally very involved in x-to-generic translations.) It's easier to justify adding yet-another construct to Generic when there is a concrete complain.

We have no formalization of Generic and I don't think we need to (or want to) have it right now, but I believe we could provide a sensible semantics for `(x?).y`, e.g. `x?` could convert a possibly null value into an `option` value, and `.` could be overload to work with `option` values. As Yoann said, for matching purposes this works perfectly well, and our current dataflow analyses don't need to know about `?.` either.

This translation lets you match both `$X?.$Y` and `$X.$Y` against `x?.y` which seems useful. The only issue is that `$X` would match `x?` which is a Generic thing. Is that a big issue?

**Comment by Sjord** [comment]
[fix(c#): use elvis operator to prevent exponential trees by brandonspark · Pull Request #8162 · returntocorp/semgrep](https://github.com/returntocorp/semgrep/pull/8162)

**Comment by IagoAbal** [comment]
Closed by #8162.
